### DISCUSSION TITLE
[A-L] (10 settembre 2024) Esercizio 4

### DISCUSSION THREAD

**Discussion by FeddyLix17** [sapienzastudentsnetwork/progettazione-di-algoritmi#117] [Esami A-L]
Sia $\Large A[1, \dots, n]$ un array di n elementi ordinato (quindi $\Large A[1] \leq A[2] \leq A[2] \leq \dots \leq A[n]$) dove ogni valore è presente esattamente 2 volte tranne un valore che è presente una sola volta.

Necessariamente il valore n deve essere dispari.

Dare il pseudocodice per un algoritmo con complessità O(log n) che prenda in input l'array A e restituisca il valore x che è presente una sola volta.

Per esempio, se $\Large A = [1,\ 1,\ 3,\ 3,\ 5,\ 6,\ 6,\ 9,\ 9,\ 10,\ 10]$ l'algoritmo restituisce "5".

**Comment by daromacarlo** [comment]
![immagine](https://github.com/user-attachments/assets/1a79e0c6-87c7-4f0f-a591-03057e549717)

**Comment by niccolo-fato** [comment]
![photo_2025-05-24_21-21-01](https://github.com/user-attachments/assets/5c07ff96-35e8-4024-a350-c22403a6a2fe)

**Comment by cek753** [comment]
```
algo(A):
m = ceil(len(A)/2)
if A[m] == A[m-1]:
    if m%2==0: return algo(A[m+1:len(A)]
    else return algo(A[0,m-1])
if A[m]!=A[m-1] and A[m]!=A[m-1]:
    return A[m]
else 
    if m%2==0: return algo(A[m+2:len(A)])
    else 
         return algo(A[0:m])
if m == 1: return A[0]
```

**Reply by Marica14** [reply]
In base al linguaggio di programmazione e a come gestisce lo slicing questa soluzione potrebbe essere in O(n). Andrebbe implementata passando gli indici nella funzione ricorsiva.
